system_zero/src/lookup.rs

```diff
@@ -0,0 +1,114 @@
+//! Implementation of the Halo2 lookup argument.
+//!
+//! References:
+//! - https://zcash.github.io/halo2/design/proving-system/lookup.html
+//! - https://www.youtube.com/watch?v=YlTt12s7vGE&t=5237s
+
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::{Field, PrimeField64};
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use starky::vars::StarkEvaluationTargets;
+use starky::vars::StarkEvaluationVars;
+
+use crate::public_input_layout::NUM_PUBLIC_INPUTS;
+use crate::registers::lookup::*;
+use crate::registers::NUM_COLUMNS;
+use crate::util::{create_hash_bag, flatten_hash_bag};
+
+pub(crate) fn generate_lookups<F: PrimeField64>(trace_cols: &mut [Vec<F>]) {
+    for i in 0..NUM_LOOKUPS {
+        let inputs = &trace_cols[col_input(i)];
+        let table = &trace_cols[col_table(i)];
+        let (permuted_inputs, permuted_table) = permuted_cols(inputs, table);
+        trace_cols[col_permuted_input(i)] = permuted_inputs;
+        trace_cols[col_permuted_table(i)] = permuted_table;
+    }
+}
+
+/// Given an input column and a table column, generate the permuted input and permuted table columns
+/// used in the Halo2 permutation argument.
+pub(crate) fn permuted_cols<F: PrimeField64>(inputs: &[F], table: &[F]) -> (Vec<F>, Vec<F>) {
+    let n = inputs.len();
+
+    // In the permuted inputs, copies of the same value must be grouped together. We accomplish this
+    // by building a hash-based bag (aka multiset) from the inputs, then flattening it.
+    let input_bag = create_hash_bag(inputs);
+    let permuted_inputs = flatten_hash_bag(&input_bag);
+
+    // Enumerate tables values that do not appear in the input list.
+    let mut unused_table_vals = table.iter().filter(|v| !input_bag.contains_key(v)).copied();
+
+    // Build the permuted table while enumerating permuted inputs. If a permuted input is a repeat,
+    // we place an unused table value, otherwise we place the permuted input value.
+    let mut permuted_table = Vec::with_capacity(n);
+    permuted_table.push(permuted_inputs[0]);
+    for i in 1..n {
+        let is_repeat = permuted_inputs[i] == permuted_inputs[i - 1];
+        permuted_table.push(if is_repeat {
+            unused_table_vals
+                .next()
+                .expect("No more unused table values; this should never happen")
+        } else {
+            permuted_inputs[i]
+        });
+    }
```

Comment:
So I was wondering if there's a way to make this more efficient using sorting instead of hash maps, and I came up with this (works on `usize` for now, and assumes `table` is sorted):
```rust
let mut permuted_inputs = inputs.clone();
permuted_inputs.sort_unstable();
let mut unused_table_inds = Vec::with_capacity(n);
let mut unused_table_vals = Vec::with_capacity(n);
let mut permuted_table = vec![0; n];
let mut i = 0;
let mut j = 0;
while (j < n) && (i < n) {
    match permuted_inputs[i].cmp(&table[j]) {
        Ordering::Greater => {
            unused_table_vals.push(table[j]);
            j += 1;
        }
        Ordering::Less => {
            if let Some(x) = unused_table_vals.pop() {
                permuted_table[i] = x;
            } else {
                unused_table_inds.push(i);
            }
            i += 1;
        }
        Ordering::Equal => {
            permuted_table[i] = table[j];
            i += 1;
            j += 1;
        }
    }
}
for jj in j..n {
    unused_table_vals.push(table[jj]);
}
for ii in i..n {
    unused_table_inds.push(ii);
}
for (ind, val) in unused_table_inds.into_iter().zip(unused_table_vals) {
    permuted_table[ind] = val;
}
```
In my benchmarks, it's 3x faster than using hash maps with `n = 2^20` (30ms vs 100ms). I don't know if this will be a bottleneck, so not sure if it's worth the added complexity.

Reply:
Interesting! I thought sorting might be a little faster but didn't expect the gap to be that significant. Did your `Ord` impl use `to_canonical`, similar to the `Hash` impl?

Reply:
So I didn't benchmark with field elements, just with `usize`s for now. But yes, to implement `Ord` on field elements I would probably compare the `to_canonical` calls. I wouldn't expect this to change the benchmark results, since the `Hash` impl also calls `to_canonical`.

Reply:
Oh got it. It's an interesting problem and seems like the costs are nontrivial, so I'll try adding a benchmark with field elements.

I think the sort method may call cmp more than we currently call hash, so potentially more `to_canonical` calls, but we could pre-canonicalize to avoid that.

I think it would be nice to support tables that aren't initially ordered, though I guess we could always pass a flag to indicate if they're already ordered.

Reply:
> I think the sort method may call cmp more than we currently call hash, so potentially more `to_canonical` calls, but we could pre-canonicalize to avoid that.

Yeah we could canonicalize the inputs and table (maybe with a new `fn canonicalize(&mut self)` function on `Field`) and then do comparison with `to_noncanonical` calls.

> I think it would be nice to support tables that aren't initially ordered, though I guess we could always pass a flag to indicate if they're already ordered.

Good point, I think in most cases we can make sure the table is sorted but I agree it makes sense to also support unsorted tables. 
Btw in my benchmark, sorting the inputs takes 75% of the time for the whole snippet (somewhat expected since the other operations are linear), so if we have to sort the table too, using sorting becomes only 2x faster than hash maps.

Reply:
I've pushed my benchmarks on the `lookups_hashmap_vs_sorting` branch if you want to have a look. A few things to note:

- I've changed the benchmarks to use field elements instead of `usize`s. The result is that using sorting now takes 55ms vs 100ms with hash maps. This is without pre-canonicalization though.
- I've benchmarked what happens when the table is the one for 16-bit range checks, i.e. `t = list(range(2**16)) + [2**16-1 for _ in range(2**16, n)]`, and interestingly sorting is way faster, taking only 10ms. I suspect it's because Rust uses some kind of linear counting sort when all values are small. 
- Again using the 16-bit range check table, the hash map based method fails. I think in general the method is buggy when the table has repeated elements. I'll add a PR comment on the relevant lines.

Reply:
Good point about duplicate table elements, I forgot we needed to support that.

I pushed my benchmark also; I read through yours and I think our methods were similar, except that I was pre-canonicalizing, and including the time to sort the table (though that's debatable since it'd be possible to skip it in our use case).

It seems sorting is slightly faster even when sorting a table of random elements, so it seems like the better option.

Reply:
Sounds good 👍 
I've tried your benchmark (modified to avoid the duplicate elements bug and assuming the table is sorted) and I get 7.7ms for sorting vs 26ms for hash maps with size 2^18. With the 16-bit range check table, I get 2.3ms vs 17ms.